src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace ContaoCommunityAlliance\Composer\Plugin\Environment;
+
+use Composer\Composer;
+use Composer\Package\RootPackageInterface;
+
+class ContaoEnvironmentFactory
+{
+
+    /**
+     * @param Composer $composer
+     *
+     * @return ContaoEnvironmentInterface
+     */
+    public function create(Composer $composer)
+    {
+        $rootDir = $this->findRoot($composer->getPackage());
+
+        if ($this->isContao3($rootDir)) {
+            return new Contao3Environment($rootDir);
+        }
+
+        if ($this->isContao4($rootDir)) {
+            return new Contao4Environment($rootDir, $composer);
+        }
+
+        throw new UnknownEnvironmentException('Contao installation was not found.');
+    }
+
+    private function isContao3($rootDir)
+    {
+        return is_file($rootDir . '/system/config/constants.php');
+    }
+
+    private function isContao4($rootDir)
+    {
+        return is_file($rootDir . '/app/console');
```

Comment:
That's not gonna work. First of all, the `uploadPath` is not set in `parameters.yml`. It can be set in `app/config.yml`, but it can also be set anywhere and imported from anywhere. And the default value is set by our bundle. The only really working thing is checking the container.

We might be able to use our own entry point or our own way of booting the kernel, but I don't know how reliable that would be (e.g. someone can "theoretically" move the `AppKernel.php` to any other place).